frontend/lib/resolution/resolution-queries.cpp

```diff
@@ -1853,80 +1853,123 @@ QualifiedType getInstantiationType(Context* context,
+
+  auto& allElements = enumElementsForEnum(context, decl);
+  size_t myIdx = 0;
+  auto enumType = initialTypeForTypeDecl(context, declId);
+  for (size_t i = 0; i < allElements.size(); i++) {
+    if (allElements[i]->id() == elementId) {
+      myIdx = i;
+      break;
+    }
 
-  // First collect all the values, no matter what types they are.
-  using ValueVector = std::vector<std::tuple<QualifiedType,
-                                             RequiredSignedness,
-                                             const EnumElement*>>;
-  ValueVector valuesAndAsts;
-  for (auto elem : enumNode->enumElements()) {
+    // The resolver will try using "local results" (i.e., consult
+    // byPostorder) for the enum elements, so give them the type they ought
+    // to have.
+    auto paramVal =
+      EnumParam::EnumValue(allElements[i]->id(), allElements[i]->name().str());
+    auto param =
+        QualifiedType(QualifiedType::PARAM, enumType,
+                      EnumParam::get(context, paramVal));
+    byPostorder.byAst(allElements[i]).setType(param);
+  }
+
+  // Found an initialization expression; use its type.
+  CHPL_ASSERT(elem);
+  optional<QualifiedType> value = empty;
+  if (elem->initExpression()) {
     elem->traverse(res);
-    QualifiedType value = {};
-
-    // Found an initialization expression; use its type.
-    if (elem->initExpression()) {
-      auto qt = byPostorder.byAst(elem->initExpression()).type();
-      auto type = qt.type();
-
-      if (qt.isErroneousType() || qt.isGenericOrUnknown()) {
-        // Don't propagate errors if they're unrelated; leave value unknown.
-      } else if (!type->isIntType() && !type->isUintType()) {
-        value = CHPL_TYPE_ERROR(context, EnumInitializerNotInteger, elem, qt);
-      } else if (!qt.isParam()) {
-        value = CHPL_TYPE_ERROR(context, EnumInitializerNotParam, elem, qt);
-      } else {
-        value = qt;
-      }
+    auto qt = byPostorder.byAst(elem->initExpression()).type();
+    auto type = qt.type();
+
+    if (qt.isErroneousType() || qt.isGenericOrUnknown()) {
+      // Don't propagate errors if they're unrelated; leave value unknown.
+    } else if (!type->isIntType() && !type->isUintType()) {
+      value = CHPL_TYPE_ERROR(context, EnumInitializerNotInteger, elem, qt);
+    } else if (!qt.isParam()) {
+      value = CHPL_TYPE_ERROR(context, EnumInitializerNotParam, elem, qt);
+    } else {
+      value = qt;
+    }
+  } else {
+    if (myIdx == 0) {
+      // we're the first value, without an init expression we're abstract.
+      value = empty;
     } else {
-      if (valuesAndAsts.empty() || std::get<2>(valuesAndAsts.back()) == nullptr) {
-        // we're either the first value, or all the previous values have
-        // been abstract. We're abstract too -- encode this using a 'null'
-        // elem.
-        elem = nullptr;
+      auto& prevValue = initialNumericValueOfEnumElement(context, allElements[myIdx - 1]->id());
```

Comment:
Does ``enumElementsForEnum`` exist to support this kind of convenient "get previous" capability? I think we could write this without the ``std::vector`` by just iterating over ``enumElements()`` and storing the previous element in a pointer outside the loop. Thoughts?